### PR TITLE
Add lmfit and emcee

### PR DESCRIPTION
**Pull request by ericdill**
Waiting on:

- [x] Confirmation of lmfit's licence: `{'license_family: BSD, license: lmfit}`
- [x] @dfm would like to be a maintainer

### CODE REVIEW

**Comment by ericdill** [comment]
@newville I am adding lmfit to conda-forge for linux, mac and windows. This will allow users to easily install this with [conda] (http://conda.pydata.org/docs/) (conda install -c conda-forge lmfit). The maintenance burden here is light as all testing and releases are built and deployed automatically with CI. Changes to how it is being built can be controlled by changing the recipe which will be located at https://github.com/conda-forge/lmfit-feedstock shortly after this PR is finished and merged. If you would like to be added as a maintainer (to make new releases for instance), I can happily add you to the maintainers list. If not, that is totally fine too!  Is there anyone else that you think would want to be pinged about maintaining this build?

**Comment by newville** [comment]
@ericdill  That sounds great to me, and I'd be happy to be added as a maintainer.    There is a simple conda recipe to start from at https://github.com/lmfit/lmfit-py/tree/master/conda-recipe

**Comment by ericdill** [comment]
@newville great, I'll add you.  I do have a question for you about the dependencies. 

To use *all* the features of lmfit you need these packages (in addition to the listed numpy/scipy):

- pandas (`lmfit/minimizer.py` and `lmfit/model.py`)
- matplotlib (`lmfit/model.py`)
- emcee (`lmfit/minimizer.py`)
- ipywidgets and IPython (`lmfit/ui/`)

These are optional packages since you can certainly use lmfit for fitting without these.  Though since conda makes it easy to install dependencies (and since it does not have the notion of "optional" packages), I am not sure what the community would expect when you `conda install lmfit -c conda-forge`.  Is it expected that it installs the same packages as `pip install lmfit`? Or would it be better to include some or all of the packages listed above in addition to `numpy` and `scipy`?

File: recipes/emcee/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{% set version = "2.1.0" %}
+
+package:
+  name: emcee
+  version: {{ version }}
+
+source:
+  fn: emcee-{{ version }}.tar.gz
+  url: https://pypi.python.org/packages/de/00/2358f12c98fa74ab58d78e6a4a4f9a8152fadf6ade2d809d98d771dcc31b/emcee-{{ version }}.tar.gz
```
Comment (ocefpaf) [inline recipes/emcee/meta.yaml]: Can you change that to https://pypi.io/packages/source/e/emcee/emcee-2.1.0.tar.gz ?

File: recipes/emcee/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  md5: c6b6fad05c824d40671d4a4fc58dfff7 
+
+build:
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record=record.txt
+
+requirements:
+  build:
+    - python
+    - numpy
```
Comment (ocefpaf) [inline recipes/emcee/meta.yaml]: `emcee` does not depend on numpy's ABI, right?

File: recipes/lmfit/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  requires:
+    - nose
+  commands:
+    - nosetests -sv {{ environ.SRC_DIR }}
+  imports:
+    - lmfit
+
+about:
+  home: http://lmfit.github.io/lmfit-py/
+  license: BSD
```
Comment (ocefpaf) [inline recipes/lmfit/meta.yaml]: Which version of BSD? `BSD 2-Clause` or `BSD 3-Clause`

File: recipes/emcee/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  imports:
+    - emcee
+  requires:
+    - nose
+  commands: 
+    - nosetests {{ environ.SRC_DIR }}/emcee -sv
+
+about:
+  home: http://danfm.ca/emcee/
+  license: GNU General Public License (GPL)
```
Comment (ocefpaf) [inline recipes/emcee/meta.yaml]: Do you know which version?

**Comment by ocefpaf** [comment]
Minor comments and this is good to go!

**Comment by newville** [comment]
@ericdill (Sorry for delay).  My view for dependencies for lmfit's conda recipe are:

 - pandas:  No.  lmfit supports using pandas dataframes, so that if someone is using pandas they can send dataframes to lmfit without having to convert them to ndarrays.  But, that means they would have pandas installed.  lmfit does not use any part of pandas.

- ipython: No.  lmfit has code to support making custom ipython widgets.  As with pandas, someone would have ipython (and ipywidgets) installed to make use of this.

- matplotlib: No.  Well, matplotlib should be part of the core ContinuumIO stack. For lmfit, there are convenience functions for plotting results from Model fitting, but this is not capabilities that are central to use of lmfit.  If someone tries this without matplotlib, they'll get an error but it's clear they have to install matplotlib.

- emcee: Yes.  lmfit can do lots of core work without emcee, but some functionality requires emcee.  If there is already a conda recipe, it should be a dependency.

Thanks!

**Comment by ericdill** [comment]
@newville Thanks for the feedback. I'll remove `pandas`, `ipython`, `ipywidgets` and `matplotlib` from the lmfit dependencies in the `meta.yaml`.  `emcee` is part of this PR too, so it will be part of the `lmfit` dependencies

File: recipes/lmfit/meta.yaml (hunk 3 above)
Comment (ericdill) [inline recipes/lmfit/meta.yaml]: I really have no idea.  The [setup.py] (https://github.com/lmfit/lmfit-py/blob/master/setup.py#L40) of lmfit just says `BSD` but the [`LICENCE`] (https://github.com/lmfit/lmfit-py/blob/master/LICENSE) file is not BSD AFAICT.  I'm not actually sure what that license is in the license file. Perhaps @newville can provide some guidance here?

File: recipes/emcee/meta.yaml (hunk 2 above)
Comment (ericdill) [inline recipes/emcee/meta.yaml]: Correct. The `setup.py` of emcee has `install_requires=["numpy"],` which, AFAICT means that it has to go in the build requirements or I get an error with text similar to "setuptools downloading is disabled during  the build phase"

File: recipes/emcee/meta.yaml (hunk 4 above)
Comment (ericdill) [inline recipes/emcee/meta.yaml]: Oh it actually has been relicensed under MIT. `conda skeleton pypi emcee` thought it was GPL. I'll update

File: recipes/emcee/meta.yaml (hunk 1 above)
Comment (ericdill) [inline recipes/emcee/meta.yaml]: done

**Comment by ericdill** [comment]
@dfm I am adding emcee to conda-forge for linux, mac and windows. This will allow users to easily install this with [conda] (http://conda.pydata.org/docs/) (conda install -c conda-forge lmfit). The maintenance burden here is light as all testing and releases are built and deployed automatically with CI. Changes to how it is being built can be controlled by changing the recipe which will be located at https://github.com/conda-forge/emcee-feedstock shortly after this PR is finished and merged. If you would like to be added as a maintainer (to make new releases for instance), I can happily add you to the maintainers list. If not, that is totally fine too!  Is there anyone else that you think would want to be pinged about maintaining this build?

**Comment by ericdill** [comment]
@ocefpaf Thanks for the review.  There are two more things to wait for before this is merged.

- [ ] Confirmation of lmfit's licence from @newville 
- [ ] Wait (a day? two?) to hear from @dfm if they want to be listed as a maintainer for this package.  I suppose that's not super necessary since I can always add them in a PR against the feedstock repo...

File: recipes/emcee/meta.yaml (hunk 4 above)
Comment (jakirkham) [inline recipes/emcee/meta.yaml]: That probably means that in their `setup` function they still reference GPL. Maybe worth letting them know.

File: recipes/emcee/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  summary: 'Kick ass affine-invariant ensemble MCMC sampling'
```
Comment (jakirkham) [inline recipes/emcee/meta.yaml]: Should we be PG-13? 😉

Comment (ericdill) [inline recipes/emcee/meta.yaml]: Hey, it's [the description in their setup.py] (https://github.com/dfm/emcee/blob/master/setup.py#L49)! Not sure it's our place to edit/censor?

File: recipes/emcee/meta.yaml (hunk 4 above)
Comment (ericdill) [inline recipes/emcee/meta.yaml]: The license in [their setup.py says MIT] (https://github.com/dfm/emcee/blob/master/setup.py#L48).  I am not sure what commit corresponds to the 2.1.0 release though, since that is not tagged on github

File: recipes/lmfit/meta.yaml (hunk 3 above)
Comment (jakirkham) [inline recipes/lmfit/meta.yaml]: Maybe MIT? That's what they say [here]( https://lmfit.github.io/lmfit-py/ ). Though the language in that [`LICENSE`]( https://github.com/lmfit/lmfit-py/blob/c31906ec9e7a67d367c869f1d7f107f31dc0c123/LICENSE ) file doesn't look like MIT.

File: recipes/lmfit/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  summary: >
+      Non-Linear Least Squares Minimization, with flexible Parameter
+      settings, based on scipy.optimize.leastsq, and with many additional
+      classes and methods for curve fitting http:/lmfit.github.io/lmfit-py/
+
+extra:
+  recipe-maintainers:
```
Comment (jakirkham) [inline recipes/lmfit/meta.yaml]: @danielballan, any interest in being added as a maintainer here? Seems like this comes from your org. 😉

Comment (ericdill) [inline recipes/lmfit/meta.yaml]: dan is on his honeymoon right now, which is why I haven't pinged him here yet.  (He is also a co-worker at my day job so I can harass him about it when he gets back :-D)

File: recipes/emcee/meta.yaml (hunk 5 above)
Comment (jakirkham) [inline recipes/emcee/meta.yaml]: I was partially joking/teasing you. 😆

Personally I am happy to have free speech, but maybe it is a point others would want to consider seriously. As far as that goes, I at least want to raise it and have everyone tell me it's fine. 😉

File: recipes/lmfit/meta.yaml (hunk 6 above)
Comment (jakirkham) [inline recipes/lmfit/meta.yaml]: Well, congrats Dan! Sorry if I distracted you.

Sounds good. I'll leave the harassing to you. 😆

**Comment by newville** [comment]
@ericdill   I don't understand what could possibly be unclear about LMFIT's license.  It's basically BSD.  It may not be a verbatim copy of the many variants of BSD, but it does include the non-endorsment clause.  Arguing over whether a license is MIT vs the BSD variants is really picking nits.   There is a license, it's permissive, it's not GPL, and it's clear.  Move on to something productive, please.

**Comment by ericdill** [comment]
@newville that feels a little hostile. I was merely trying to understand.

**Comment by newville** [comment]
@ericdill  don't mean to be hostile, but I really do not want to be sucked in to further discussion about licensing.  The license is clear.  Whether it is BSD or MIT is so unimportant that I feel like it's a waste of my time even to write this response.  Really, not hostile about making a conda recipe, but I'm finished discussing the license.

File: recipes/lmfit/meta.yaml (hunk 3 above)
Comment (jakirkham) [inline recipes/lmfit/meta.yaml]: Alright, let's resolve this. Some options. Either seem fine to me.

1. `BSD-like`. [Example]( https://github.com/conda-forge/pyopengl-feedstock/blob/9dbfdeba92891e1b73112786d44c76013ae6f534/recipe/meta.yaml#L30 ).
2. Call it the `lmfit` license. [Example]( https://github.com/conda-forge/hdf5-feedstock/blob/38ee94d15eff502b4c387def4ae87c759978aee3/recipe/meta.yaml#L93-L94 ).

Comment (ericdill) [inline recipes/lmfit/meta.yaml]: I like option 2. Thanks for the suggestion.

**Comment by ericdill** [comment]
The tests for lmfit 0.9.3 just started failing on circleci.  Looking at the packages that are getting installed now, something funny might be going on. Seems to be installing the blas_openblas *and* mkl...?

```

The following NEW packages will be INSTALLED:

    blas:            1.1-openblas                  conda-forge                          
    ca-certificates: 2016.2.28-0                   conda-forge                          
    emcee:           2.1.0-py27_0                  file:///opt/conda/conda-bld/linux-64/
    libgcc:          5.2.0-0                       defaults                             
    libgfortran:     3.0.0-1                       defaults                             
    lmfit:           0.9.3-py27_0                  file:///opt/conda/conda-bld/linux-64/
    mkl:             11.3.3-0                      defaults                             
    nose:            1.3.7-py27_0                  defaults                             
    numpy:           1.11.0-py27_blas_openblas_100 conda-forge                           [blas_openblas]
    openblas:        0.2.18-1                      conda-forge                          
    openssl:         1.0.2h-0                      conda-forge                          
    pip:             8.1.1-py27_1                  defaults                             
    python:          2.7.11-0                      defaults                             
    readline:        6.2-2                         defaults                             
    scipy:           0.17.1-np111py27_0            defaults                             
    setuptools:      20.7.0-py27_0                 defaults                             
    sqlite:          3.11.1-0                      conda-forge                          
    tk:              8.5.19-0                      conda-forge                          
    wheel:           0.29.0-py27_0                 defaults                             
    zlib:            1.2.8-1                       conda-forge                          
===== testing package: lmfit-0.9.3-py27_0 =====
import: 'lmfit'
===== lmfit-0.9.3-py27_0 OK =====
+ nosetests -sv /opt/conda/conda-bld/work/lmfit-py-0.9.3
/opt/conda/conda-bld/work/lmfit-py-0.9.3/lmfit/ui/__init__.py:17: UserWarning: lmfit.Fitter will use basic mode, not IPython: need matplotlib
  warnings.warn("lmfit.Fitter will use basic mode, not IPython: need matplotlib")
test_1variable.test_1var ... Intel MKL FATAL ERROR: Cannot load libmkl_avx.so or libmkl_def.so.
TESTS FAILED: lmfit-0.9.3-py27_0
```

File: recipes/lmfit/meta.yaml (hunk 3 above)
Comment (ocefpaf) [inline recipes/lmfit/meta.yaml]: > I like option 2. Thanks for the suggestion.

+1 for option 2. Sorry for the confusion Eric!

**Comment by jakirkham** [comment]
Please see this issue ( https://github.com/conda-forge/staged-recipes/issues/628 ), @ericdill.

**Comment by ericdill** [comment]
@jakirkham Do I understand the resolution of this issue is going to be to package scipy in conda-forge?

**Comment by jakirkham** [comment]
> @jakirkham Do I understand the resolution of this issue is going to be to package scipy in conda-forge?

Depends on your timeline. My goal is to try and get it in by the end of the day. If that works for you, yes. If not, pinning `mkl` and shipping it might be a faster (and already acceptable) resolution. Thoughts?

**Comment by jakirkham** [comment]
Should add that Windows builds of NumPy and SciPy are blocked until we have a working OpenBLAS package for Window.

**Comment by ericdill** [comment]
> Depends on your timeline

Well I'm trying to move the whole nsls-ii stack over to conda-forge, so I am in no real hurry since there are a lot of packages (~50) to re-package with conda-forge!

> pinning mkl

What is involved with this?  Is this an acceptable long-term solution?

**Comment by ericdill** [comment]
Oops hit submit too soon...

Is pinning mkl an acceptable long-term solution or will that change need to be reverted once we have working builds of numpy and scipy on all platforms?

**Comment by jakirkham** [comment]
> Well I'm trying to move the whole nsls-ii stack over to conda-forge, so I am in no real hurry since there are a lot of packages (~50) to re-package with conda-forge!

Know the feeling. 😄 Let me rephrase though. How soon do you need `lmfit`?

> What is involved with this?

Easy pin `mkl` to `11.3.1` in `build` and `run`. We have seen it on Linux and Windows. They may have fixed Windows, but Linux is still in trouble.

> Is pinning mkl an acceptable long-term solution or will that change need to be reverted once we have working builds of numpy and scipy on all platforms?

No. This is a workaround.

<br>
Based on your feedback, I would recommend that we hold on this until `scipy` is in conda-forge. Again this is top prio for me. It should be completable today. Does that sound reasonable to you?

**Comment by ericdill** [comment]
> Does that sound reasonable to you?

Yup! Sounds just fine.

> How soon do you need lmfit?

By mid-June I'd like to get the whole stack done.  lmfit is an upstream library for three of those packages (one of which is a metapackage), so there is no real hurry here.  If I could get this into conda-forge within ~3 weeks I'd be satisfied

**Comment by dfm** [comment]
@ericdill: Sorry about the delayed response. I'd love to be listed as a maintainer on emcee and I'll get to your pull request soon. Thanks!

**Comment by ericdill** [comment]
> Sorry about the delayed response

I only pinged you a few hours ago, so it doesn't seem like the response was that delayed! :grinning: 

> I'd love to be listed as a maintainer on emcee

Ok, I'll add you

> I'll get to your pull request soon

Great!  It'd be good to get automated testing on 3.4 and 3.5 running since conda-forge only releases those two minor version of python 3 (though it isn't too difficult to turn it on 3.2/3.3 IIUC)

File: recipes/lmfit/meta.yaml (hunk 6 above)
Comment (danielballan) [inline recipes/lmfit/meta.yaml]: Thanks! I got back last night. I'd be happy to be added as a co-maintainer.

**Comment by ericdill** [comment]
@jakirkham nice work on scipy. Problems seem to be resolved!

File: recipes/emcee/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/emcee/meta.yaml]: GitHub also shows no C or C++ code, which I find to be a good indicator too. As we are not using Cython here, that also helps to confirm this.

**Comment by jakirkham** [comment]
LGTM.

>  If I could get this into conda-forge within ~3 weeks I'd be satisfied...

How about today? 😉

**Comment by jakirkham** [comment]
Thanks @ericdill.

**Comment by ericdill** [comment]
Thanks for the comments everyone.
